### PR TITLE
Add New Test to Monitor Number of Jitted Methods

### LINKED ISSUE #88533: R2R: Add a CI or perf validation for # of methods jitted
We have recently seen more methods being JITed for simple "helloworld" application, so getting visibility and ensuring we can minimize proliferation of these via automation would be helpful. Note that we are continuously working on improving R2R handling of new functionality, but always good to track this metric.

### PR DESCRIPTION
**Pull request by ivdiazsa**
Solution of issue #88533. This PR adds a new test to the `JIT/Performace` tree to track the number of methods jitted at runtime in a predefined app.

The test's function is straightforward. It sets up an environment where we can run a dotnet app and capture the methods jitted at runtime into a file. In this PR, a _HelloWorld_ app is enough to fulfill our purposes. Then, it runs the app and if everything goes fine, it reads the file with the jitted methods written down to see how many there were. Based on this number, the test's result is defined successful or not.

For further information, the initial discussion took place in Tomas' design-only PR #90647.

### CODE REVIEW

**Comment by ivdiazsa** [comment]
cc @jkotas

File: src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            };
+
+            // Set up the environment for running the test app. We are looking
+            // to seeing how many methods were jitted at runtime. So, we ask
+            // the runtime to print them out with DOTNET_JitDisasmSummary, and
+            // write them out to a file we can parse and investigate later
+            // with DOTNET_JitStdOutFile.
+
+            startInfo.EnvironmentVariables.Add("DOTNET_JitDisasmSummary", 1);
+            startInfo.EnvironmentVariables.Add("DOTNET_TieredCompilation", 0);
```
Comment (jkotas) [inline src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:63]: Why are we disabling tiered compilation for this test?

File: src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.csproj [hunk 2]
```diff
@@ -0,0 +1,24 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Compile Include="JittedMethodsCountingTest.cs" />
+  </ItemGroup>
```
Comment (jkotas) [inline src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.csproj:10]: Are there configurations that this test needs to be disabled for? I am thinking about JIT stress, etc.

**Comment by jkotas** [comment]
cc @dotnet/jit-contrib

File: src/tests/JIT/Performance/JIT.performance.csproj [hunk 3]
```diff
@@ -1,6 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <ItemGroup>
     <MergedWrapperProjectReference Include="*/**/*.??proj" />
+    <MergedWrapperProjectReference Remove="JittedMethodsCountingTest/HelloWorld/Helloworld.csproj" />
```
Comment (trylek) [inline src/tests/JIT/Performance/JIT.performance.csproj:4]: You can fix this more cleanly by modifying the following line:

https://github.com/dotnet/runtime/blob/7a11cff914fa6fcf00eb59be33306e3826e8b958/src/tests/Directory.Build.targets#L513

as follows:

<pre>
  &lt;PropertyGroup Condition="'$(Language)' == 'C#' and $(_CLRTestNeedsToRun) and ('$(BuildAsStandalone)' == 'true' or '$(RequiresProcessIsolation)' == 'true' or '$(InMergedTestDirectory)' == 'true' or '$(IsMergedTestRunnerAssembly)' == 'true')">
</pre>

Once you do that, it should suffice to state in your ```HelloWorld.csproj``` project that it's ```&lt;CLRTestKind&gt;BuildOnly&lt;/CLRTestKind&gt;``` and it should do the same thing under the covers.

**Comment by trylek** [comment]
@jkoritzinsky - I was mistaken in my yesterday assessment, the error indeed comes from the compilation of HelloWorld as you expected, we don't check the transitive references for entry points. That ultimately simplified the change because we can just conditionally skip emitting the wrapper - we already have a few places in the tree that explicitly set ```ReferenceXUnitWrapperGenerator``` to false but that seemed a bit dirty to me so I proposed to Ivan to make a change to ```src/tests/D.B.targets``` that will suppress the generator for ```<CLRTestKind>BuildOnly</CLRTestKind>``` projects.

File: src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs (hunk 1 above)
Comment (trylek) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:63]: The HelloWorld test should be very short, perhaps we could monitor both. I agree it doesn't make much sense to run it under stress.

Comment (jkotas) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:63]: The runtime tests are run in multiple configurations (tiered compilation on/off, etc.). The test itself should not need to explicitly cover different configurations.

Comment (trylek) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:63]: Sounds reasonable, in such case it should suffice to remove this environment variable override and the child process (```corerun```) should automatically inherit the parent settings.

Comment (trylek) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs]: Sure, you should be able to achieve that by putting the property ```<CLRTestTargetUnsupported Condition="'$(TestBuildMode)' != 'crossgen2'">true</CLRTestTargetUnsupported>``` in the ```JittedMethodsCountingTest.csproj``` script. Please note that in such case you also need to mark the test as ```<RequiresProcessIsolation>true</RequiresProcessIsolation>``` as today merged wrapper generator isn't smart enough to generate a variant wrapper capable of identifying tests valid for a particular execution mode. As the number of such tests is generally low, for now we haven't tried to devise a special mechanism for them.

Comment (trylek) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs]: (Alternatively, you can just keep the setting ```DOTNET_ReadyToRun=1``` and run the test in both R2R and JITted pipeline configurations, it's probably not going to be a fundamental contributor to our total test running time. In such case you'd need to mark the ```HelloWorld``` test with the property ```<AlwaysUseCrossGen2>true</AlwaysUseCrossGen2>``` so that it always crossgens the test no matter what the ambient setting is.

File: src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            Console.WriteLine("App Exit Code: {0}", appExitCode);
+            Console.WriteLine("Jitted Methods Generated File: {0}", jitOutputFile);
+        }
+
+        return appExitCode;
+    }
+
+    private static int GetNumberOfJittedMethods(string jitOutputFile)
+    {
+        string[] lines = File.ReadLines(jitOutputFile).ToArray();
```
Comment (jkotas) [inline src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:81]:
```suggestion
        string[] lines = File.ReadAllLines(jitOutputFile);
```

File: src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs (hunk 1 above)
Comment (ivdiazsa) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs]: I was thinking on this yesterday at home, and I came up with another alternative I think might be cleaner. How about checking for `DOTNET_ReadyToRun` at the beginning of the test, and if it's not enabled, end it right there returning _100_? Also, print out a message saying the test is not going to run because it's not supported outside R2R-enabled contexts. What are your thoughts on this Tomas?

Comment (trylek) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs]: That sounds good to me, in fact it's likely going to be faster in the lab because this way your test can run in process of the merged wrapper and it will just finish immediately in the absence of ```DOTNET_ReadyToRun```. Please just remember that the default is 1 so that you need to test whether it's been explicitly overridden to 0.

Comment (ivdiazsa) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs]: Sounds good. I was thinking maybe check whether it's different from 1 instead?

Comment (trylek) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs]: Whatever you prefer, just keep in mind that "" (the variable not being set) means "1", not "0".

Comment (ivdiazsa) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs]: Alright got it. I'll then make the test run only if `DOTNET_ReadyToRun` is set to "1" or not set at all, since that implicitly means "1" as well.

File: src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs [hunk 5]
```diff
@@ -9,14 +9,30 @@
...
+        // This can potentially cause issues, so we have to make sure it is
+        // created exclusively for this test's purpose.
+        if (File.Exists(jitOutputFile))
+        {
+            File.Delete(jitOutputFile);
+        }
+
         // For adding any new apps for this test, make sure their success return
         // code is 0, so we can universally handle when they fail.
         int appResult = RunHelloWorldApp(appName, jitOutputFile);
```
Comment (trylek) [inline src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:38]: Should we include a check for the ```appResult``` value in the expression for the exit code?

File: src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs [hunk 6]
```diff
@@ -34,6 +50,12 @@ public static int TestEntryPoint()
         return jits > 0 && jits <= MAX_JITTED_METHODS_ACCEPTED ? 100 : 101;
     }
 
+    private static bool IsReadyToRunEnvSet()
+    {
+        string? dotnetR2R = Environment.GetEnvironmentVariable("DOTNET_ReadyToRun");
+        return (dotnetR2R is null || dotnetR2R == "1") ? true : false;
```
Comment (trylek) [inline src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:56]: I have a vague feeling (you probably know that better than I do) that on Linux you can end up with either ```null``` or ```empty string``` both of which should probably have the same behavior i.o.w. we should probably use ```string.IsNullOrEmpty``` (or return to my original lame suggestion that inequality to ```"0"``` is good enough).

**Review by trylek — APPROVED** [review]
LGTM, thanks Ivan! I have shared just a couple of non-blocking nitpicks, overall your change looks great to me provided it actually passes in the lab.

File: src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs (hunk 6 above)
Comment (ivdiazsa) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:56]: It also returns `null` on Linux. However, since any kind of empty is equivalent to "1" in this case, we can play it safer by going the `string.IsNullOrEmpty()` route instead :+1:

File: src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs (hunk 5 above)
Comment (ivdiazsa) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:38]: I don't think that would be necessary. We're already marking the test as failed and returning 101 if the app's exit code is different from 0 (i.e. something went wrong there).

File: src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using Xunit;
+
+public class JittedMethodsCountingTest
+{
+    private const int MAX_JITTED_METHODS_ACCEPTED = 120;
```
Comment (AndyAyersMS) [inline src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: 120 seems like a pretty high number to allow -- do we want to set this lower?

File: src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.csproj (hunk 2 above)
Comment (AndyAyersMS) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.csproj:10]: Yeah, adding `<JitOptimizationSensitive>true</JitOptimizationSensitive>` to the `PropertyGroup` seems like a good idea.

File: src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs (hunk 7 above)
Comment (EgorBo) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: Empty app on Linux/macOS x64/arm64 is expected to produce just [1 jit compilation](https://github.com/dotnet/runtime/pull/90412#issuecomment-1679822542) (for Main) 🙂 
On windows there can be more due to background ETW sessions, but not too much (like 16 in total or so)

For Console.WriteLine it's tricky because `System.Console` is not prejitted in dotnet/runtime, hence, all methods in it are always jitted

Comment (ivdiazsa) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: I must admit I'm not too familiarized with the expected jitted methods counts. What would be a good expected range for a HelloWorld app like the one added in this project?

Comment (trylek) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: I think we just need a baseline and go from there. I recall that in my experiments with startup of the "dotnet new webapi" I observed about 50 methods being jitted at runtime when compiled as composite with the MIBC optimization data i.o.w. in more or less "the best possible configuration".

Comment (trylek) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: (Without the MIBC data it was almost 200 IIRC.)

Comment (jkotas) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: This test is only protecting against something going completely wrong with crossgen of System.Private.CoreLib. It does not run on actual shipping bits as Egor mentioned and it runs in unstable environment (background ETW sessions) that needs to allow for margin of error.

Is this test going to catch the two issues that sneaked through that motivated us to add it?

Comment (jkotas) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: > I think we just need a baseline and go from there.

The construction of this test and where we have added it does not really allow us to go from here.

Comment (trylek) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: Hmm, I don't understand, can you please clarify? I would assume that, once this test has been merged in, anyone is welcome to establish a tighter upper bound for the number of jitted methods provided their lab testing passes. If you're alluding to the fact that the test is supposed to run in various configurations (with tiered compilation on vs. off, on different OSes and architectures), I guess it's up to further discussion if we're fine with something like a "greatest common divisor" in the sense of maximum methods jitted in all configurations or if we need something like a table with the expected method counts per various build platforms / execution modes.

Comment (trylek) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: FWIW I just tried with "dotnet new console" compiled using ".NET 8.0 preview 7 SDK" in release mode (Windows x64) with ```PublishReadyToRun=true``` and ```PublishReadyToRunComposite=true``` (without the MIBC optimization data) and I'm seeing 109 methods JITted at runtime.

Comment (trylek) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: If we come to the conclusion that ```Console.WriteLine``` is skewing the number of jitted methods in an undesirable manner, I guess it should be no big deal for Ivan to delete the one line from his test app :-); in a way it may be an even better approximation of a really empty app doing just the runtime initialization and shutdown.

Comment (jkotas) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: > I don't understand, can you please clarify?

We need to be clear about the purpose of this test.

As it is setup, this test protects against something going very wrong with crossgen of System.Private.CoreLib. Is this the intended long-term purpose of this test? If not, what is it instead?

I do not think we should be treating this test as a perf test. It does not test the actual product bits and it would not scale well to larger apps like `dotnet webapi`. Also, we should not be optimizing number of JITed methods without looking at the wall clock time that these methods took to JIT. The primary metric that we care about is the startup time and the time that the methods participating in startup took to JIT. The number of the JITed methods is a secondary metric and it can be easily skewed (e.g. merging two small methods into one big method). There are classic anecdotes in MS perf community about how focus on secondary metric resulted into the primary metric regressing, we do not want to the next one.

I think we should follow the same approach for startup as we do with other benchmarks and allow the perflab to be the primary place that it is measured and tracked. I would like to see the nice graphs that show how the time spend JITing methods and the number of JITed methods in `dotnet new console` or `dotnet new webapi` startup kept changing over the course of the release. Does something like that exist already?

Note that the existing tests under `src\tests\JIT\Performance` where this test is being added are not run as a performance benchmarks. These tests exist in the test tree to ensure they are working and to allow JIT team infra see the asm diffs for them. The equivalent benchmark versions of these tests live in perf benchmarks repo under https://github.com/dotnet/performance/tree/main/src/benchmarks/micro/runtime and these copies run in the perflab.

Comment (jkotas) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: > We want to monitor the number of jitted methods. Nothing more, nothing less. 

This is not clear explanation for me. It explains the work. It does not explain what we want to achieve.

> In that case, where would you suggest placing it?

If you want to treat it as a perf test, it should live in https://github.com/dotnet/performance along with all other runtime perf benchmarks that we have.

Comment (jkotas) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: > We can always add new apps and make it more configurable, which is the longer term goal. 

The core runtime tests are not the best place for adding end-to-end (perf) tests.

Comment (ivdiazsa) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: I understand your point, even if I do not agree with it. But I have a question. Why didn't you provide these ideas when you gave your first feedback? I find it difficult to digest to get approved only for at the end be told that basically it should be done an entirely different way. Even in an entirely different place.

Comment (mangod9) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: From my perspective the intent of this test in the runtime repo is to smoke test to ensure that R2R of SPCL is not completely broken. Currently we would only notice such a break when microbenchmarks or TechEmpower flag a large startup regression. My thinking would be that we are far more lenient with the allowed number of JITted methods.

Comment (mangod9) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: as mentioned here: https://github.com/dotnet/runtime/issues/88533#issuecomment-1626319472.

Comment (EgorBo) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: We already have simliar smoke-tests, right? like that NAOT test where we check that an empty app's size is not terribly big in dotnet/runtime repo - is it real-world? No, but better than nothing and, afair, it already helped once. I agree that this test as is won't help us to find cases when something goes wrong but not terribly (e.g. 100 methods regressed). For that we still need to apply efforts elsewhere: I'd want to have a separate graph for "methods jitted" counter at https://aka.ms/aspnet/benchmarks so once startup regresses again I will be able at least quickly validate that it's not because of R2R. A visualization on dotnet/performance would work too.

Comment (BruceForstall) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: > As for where the runtime repo version should be placed, considering what you explained about `src/tests/JIT/Performance`, then I agree it's not the best place. In that case, where would you suggest placing it?

I agree that "JIT/Performance" is not the right place.

How about src\tests\readytorun\...?

Comment (jkotas) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: > Why didn't you provide these ideas when you gave your first feedback? 

I have provided similar feedback in the issue discussion - look for my comments in https://github.com/dotnet/runtime/issues/88533.

> the test is ok to have IMO

Agree - this test as a smoke test that will catch things going terribly wrong. It is not a perf gate.

Comment (ivdiazsa) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: > We already have similar smoke-tests in dotnet/runtime, right? like that NAOT test where we check that an empty app's size is not terribly big in dotnet/runtime repo - is it real-world? No, but better than nothing and, afair, it already helped once. I agree that this test as is won't help us to find cases when something goes wrong but not terribly (e.g. 100 methods regressed). For that we still need to apply efforts elsewhere: I'd want to have a separate graph for "methods jitted" counter at https://aka.ms/aspnet/benchmarks so once startup regresses again I will be able at least quickly validate that it's not because of some obvious R2R issue. A visualization on dotnet/performance would work too.
> 
> tldr: the test is ok to have IMO, additional tests in dotnet/performance and PerfLab would be nice to have too (where we work with SDK)

Thanks for your feedback @EgorBo. I agree having a visual graph would certainly help us catch regressions-in-the-making sooner. Then I have a follow up question for you and @AndyAyersMS. What would be a good number of jitted methods that if exceeded, we can confirm something went terribly wrong and therefore fail this test?

As for the dashboard and graphs, I will reach out to Sebastien to get it to run periodically in TE alongside the other perf tests we already run (e.g. ASP.NET Composite Containers).

> How about src\tests\readytorun...?

I think this is a good place to put this one @BruceForstall. I'll move it and rerun the lab tests.

Comment (jkotas) [reply src/tests/JIT/Performance/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:12]: > What would be a good number of jitted methods that if exceeded, we can confirm something went terribly wrong and therefore fail this test?

It should be the lowest number that makes the test pass reliably with some margin. What is the highest number of methods that you see being JITed across different configurations? Add 20 to it and use that as the limit.

> I will reach out to Sebastien to get it to run periodically in TE alongside the other perf tests we already run (e.g. ASP.NET Composite Containers)

A trivial `Console.WriteLine` startup perf test has nothing to do with ASP.NET. It does not sound right to add it to ASP.NET perf infra for TE.

We have startup-oriented scenario perf measurements under https://github.com/dotnet/performance/tree/main/src/tools/ScenarioMeasurement. We should have a guidance for when to add a scenario measurement to this location vs. to the TE infrastructure. Please check with Sebastien and Scott Blomquist about that.

**Comment by ivdiazsa** [comment]
Hmmm seems like the test can't find the file in the lab runs. Looking into it right now.

**Comment by markples** [comment]
@ivdiazsa This sounds like the kind of thing where the specialized test logic does something in a subdirectory. Perhaps the current working directory and test binary directory don't match and the write/read of the file then don't match?

**Comment by ivdiazsa** [comment]
> @ivdiazsa This sounds like the kind of thing where the specialized test logic does something in a subdirectory. Perhaps the current working directory and test binary directory don't match and the write/read of the file then don't match?

That seems to be the case, although I'm having trouble figuring out why. I can't seem to reproduce it on my machine so it must be some Helix configuration that is messing up with the paths. I'll try using the current working directory explicitly as reference instead of assuming and hopefully that should help.

File: src/tests/readytorun/JittedMethodsCountingTest/JittedMethodsCountingTest.cs [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        // If DOTNET_ReadyToRun is disabled, then this test ought to be skipped.
+        if (!IsReadyToRunEnvSet())
+        {
+            Console.WriteLine("\nThis test is only supported in ReadyToRun scenarios."
+                              + " Skipping...\n");
+            return 100;
+        }
+
+        string appName = "HelloWorld.dll";
+        string jitOutputFile = "jits.txt";
```
Comment (trylek) [inline src/tests/readytorun/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:26]: I think you should generate a full path akin to what I did in #90647,

```
      string testAppFolder = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);

        string appName = Path.Combine(testAppFolder, "HelloWorld.dll");
        string jitOutputFile = Path.Combine(testAppFolder, "jits.txt");
```

The scripts may be tampering with the current directory.

Comment (ivdiazsa) [reply src/tests/readytorun/JittedMethodsCountingTest/JittedMethodsCountingTest.cs:26]: Thanks for the suggestion Tomas! Will try it this way.
